src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+            if (controllerData.pickIndexMeshTip) {
+                let utilitySceneNearPick = null;
+                if (this._options.useUtilityLayer && this._utilityLayerScene) {
+                    utilitySceneNearPick = this._pickWithMesh(controllerData.pickIndexMeshTip, this._utilityLayerScene, false, this.nearPickPredicate);
+                }
+                let originalSceneNearPick = this._pickWithMesh(controllerData.pickIndexMeshTip, this._scene, false, this.nearPickPredicate);
+                let pickInfo = accuratePickInfo(originalSceneNearPick, utilitySceneNearPick);
+                pick = populateNearInteractionInfo(pickInfo);
+            }
+
+            // near interaction grab
+            if (controllerData.grabRay) {
+                let utilitySceneNearGrab = null;
+                if (this._utilityLayerScene) {
+                    utilitySceneNearGrab = this._utilityLayerScene.pickWithRay(controllerData.grabRay, this.nearGrabPredicate);
+                }
+                let originalSceneNearGrab = this._scene.pickWithRay(controllerData.grabRay, this.nearGrabPredicate);
+                pick = accuratePickInfo(originalSceneNearGrab, utilitySceneNearGrab);
+                if (pick && pick.pickedPoint && pick.hit) {
+                    controllerData.nearGrab = true;
+                    controllerData.nearInteraction = true;
+                    pick.ray = controllerData.grabRay;
+                }
+            }
+
+            // Turn on far interaction if near interaction is unsuccessful
+            if (!controllerData.nearInteraction && !hoverRange && !controllerData.nearGrabInProcess) {
+                if (this._farInteractionFeature) {
+                    this._farInteractionFeature.attach();
+                }
+            }
+            controllerData.pick = pick;
+
+            if (pick && pick.pickedPoint && pick.hit) {
+                controllerData.meshUnderPointer = pick.pickedMesh;
+            } else {
+                controllerData.meshUnderPointer = null;
+            }
+        });
+    }
+
+    private get _utilityLayerScene() {
+        return this._options.customUtilityLayerScene || UtilityLayerRenderer.DefaultUtilityLayer.utilityLayerScene;
+    }
+
+    private _attachTrackedPointerRayMode(xrController: WebXRInputSource) {
+        const controllerData = this._controllers[xrController.uniqueId];
+        const pointerEventInit: PointerEventInit = {
+            pointerId: controllerData.id,
+            pointerType: "xr",
+        };
+        controllerData.onFrameObserver = this._xrSessionManager.onXRFrameObservable.add(() => {
+            if (controllerData.pick) {
+                this._scene.simulatePointerMove(controllerData.pick, pointerEventInit);
+            }
+        });
+        if (xrController.inputSource.gamepad) {
+            const init = (motionController: WebXRAbstractMotionController) => {
+                // squeezeComponent is a new member on the controllerData

```

Comment:
This is one of the comments I added for your reference when implementing, it doesn't make much sense having it stick around